spinalcordtoolbox/reports/slice.py

```diff
@@ -444,3 +456,44 @@ def get_center(self, img_idx=-1):
         size_y = self.axial_dim(image)
         size_x = self.coronal_dim(image)
         return np.ones(dim) * size_x / 2, np.ones(dim) * size_y / 2
+
+    def mosaic(self, nb_column=0, dilation=(0, 0, 0)):
+        """Obtain matrices of the mosaics
+
+        Mosaic images are cropped based on the bounding box of the spinal cord segmentation.
+
+        :param nb_column: number of mosaic columns
+        :param dilation: a tuple containing (x,y,z) integers for dilating the bounding box used for cropping.
+        :return: tuple of numpy.ndarray containing the mosaics of each slice pixels
+        """
+        # 0. Use the segmentation image to initialize the image cropper
+        self._image_seg = self._images[-1]
+        cropper = ImageCropper()
+        cropper.get_bbox_from_mask(self._image_seg)
+
+        # 1. Compute the sizes of the patches, as well as the overall image
+        # 1a. Compute width and height of mosaic squares. (This is assumed to be a square for Axial slices.)
+        size_h = cropper.bbox.xmax - cropper.bbox.xmin + 1 + (2 * dilation[0])  # SAL -> SI axis provides height
```

Comment:
Ahhh, I see your point here.

- `ImageCropper.crop()` limits the dilation from going outside of the image dimensions:
   https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/f0a96439dbc822136b42653e803b5f89cfa3b3ed/spinalcordtoolbox/cropping.py#L48-L53
- But, `size_h`/`size_w` don't account for this limiting, so the computed dimensions can be larger than the cropped images
- And, trickily, `dilation` is only applied during the actual cropping, which makes it hard to "precompute" the correct size.

It seems like to get the correct size, we would need to perform dilation _before_ cropping, then fetch the size as some sort of `@property` of the bounding box.  But, that has its own caveats, because making `dilation` a method instead of a crop parameter means you can no longer vary the dilation across multiple crops. (Assuming this pattern is even used/wanted?)

I agree with your mention re: architecture. This messiness feels very tied to the points you raised in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4102#pullrequestreview-1427320288 regarding the structure of `ImageCropper`/`BoundingBox`.

Reply:
I think the simplest short-term fix here is to remove the optional argument `dilation=(0,0,0)` from `Sagittal.mosaic` for now, sadly. We could probably re-add it after a refactoring of the `ImageCropper` code to make it less corner-casey.

Reply:
My main reason for adding the dilation feature was just in case we wanted to crop outside the spinal cord segmentation for lesion detection purposes. (Admittedly I don't understand spinal cord lesions well, but I was wondering if maybe it would be possible for a lesion to exist outside of the detected spinal cord segmentation, and so we would want some buffer to examine the surrounding areas.)

But, I don't know if this is relevant enough to make this a necessity? So, I would want to ask @valosekj, since I imagine he has more of an understanding of lesion segmentation than I do, and can try the QC on more images.

Reply:
>  I was wondering if maybe it would be possible for a lesion to exist outside of the detected spinal cord segmentation, and so we would want some buffer to examine the surrounding areas.)

Great point! From the anatomical point of view, both multiple sclerosis (MS) and spinal cord injury (SCI) lesions should be presented within the spinal cord. But, as you correctly assume, the spinal cord segmentation algorithm (`sct_deepseg_sc`) sometimes fails at the lesion site (due to rapid change of intensities in the lesions relative to the spinal cord). In other words, the lesion is not included within the spinal cord segmentation mask. 
But, the new contrast-agnostic model works pretty well even in these cases. So I believe we can remove the optional argument `dilation=(0,0,0)` (as mentioned by @mguaypaq in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4102#discussion_r1199288297).